Add initial implementation with support for decoding CLP IR streams. (#1)
y-scope/clp-ffi-js: CMakeLists.txt

cmake_minimum_required(VERSION 3.22)

include(FetchContent)

project(ClpIrV1Decoder
    LANGUAGES CXX C
)

# Enforce "Unix Makefiles" because "Ninja" does not work with the Emscripten toolchain on CXX_20 sources.

> if(NOT CMAKE_GENERATOR MATCHES "Unix Makefiles")

Reviewer: nit: add space after `if`
```suggestion
if (NOT CMAKE_GENERATOR MATCHES "Unix Makefiles")
```

    message(FATAL_ERROR "This project is intended to be built with Unix Makefiles on a Unix system.")

> endif()

Reviewer: nit: add space after `endif`
```suggestion
endif ()
```

# Enable compile commands by default if the generator supports it.
if (NOT CMAKE_EXPORT_COMPILE_COMMANDS)
    set(CMAKE_EXPORT_COMPILE_COMMANDS ON CACHE BOOL
        "Enable/Disable output of compile commands during generation." FORCE)
endif()

# Set the default build type to Release if not specified

> if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)

Reviewer: nit: add space after `if`
```suggestion
if (NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
```

>     message(STATUS "Setting build type to 'Release' as none was specified.")

Reviewer: Let's add a variable to deduplicate (a general practice to avoid manual errors as well).
```suggestion
    set(default_build_type "Release")
    message(STATUS "No build type specified. Setting to '${default_build_type}'.")
```

>     set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Choose the type of build (Debug, Release, RelWithDebInfo, MinSizeRel)." FORCE)

Reviewer: Thanks for the intuitive prompt message. The other release types are less common and we don't have specializations for those. Maybe we can remove those for now.
```suggestion
    set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE STRING "Choose the type of build." FORCE)
```

> endif()

Reviewer: nit: add space after `endif`
```suggestion
endif ()
```

# Download and extract Boost
set(BOOST_VERSION "1.85.0" CACHE STRING "Version of Boost library to be used when compiling CLP sources.")
string(REPLACE "." "_" BOOST_ARCHIVE_VERSION_PART ${BOOST_VERSION})

> set(BOOST_SRC_DIR "${CMAKE_CURRENT_SOURCE_DIR}/build/wasm/boost" CACHE STRING

Reviewer: Let's avoid hard-coding the build path to "${CMAKE_CURRENT_SOURCE_DIR}/build" because users should be able to name their build directories however they want (e.g., CLion generates cmake-build-release / cmake-build-debug). Instead, let's use "${CMAKE_CURRENT_BINARY_DIR}" which will be the directory where the cmake command was run (in your case, still "build").
```suggestion
set(BOOST_SRC_DIR "${CMAKE_CURRENT_BINARY_DIR}/boost" CACHE STRING
```

    "Directory containing Boost sources, fetched from official release if non existent.")
if (NOT EXISTS ${BOOST_SRC_DIR})
    FetchContent_Declare(
        Boost
        URL "https://archives.boost.io/release/${BOOST_VERSION}/source/boost_${BOOST_ARCHIVE_VERSION_PART}.tar.gz"
        SOURCE_DIR ${BOOST_SRC_DIR}

>     )

Reviewer: We should add this to let cmake to override the source modification times when decompressing files from the archive, causing certain files not to recompile. A warning is shown if we don't add that.
```suggestion
        DOWNLOAD_EXTRACT_TIMESTAMP ON
    )
```

    message(STATUS "Fetching Boost from official release.")
    FetchContent_MakeAvailable(Boost)
endif()

set(BIN_NAME "ClpIrV1Decoder" CACHE STRING "Binary name for the generated .js file and .wasm file of CLP IR V1 Decoder.")
add_executable(${BIN_NAME})

target_compile_features(${BIN_NAME}
    PRIVATE
    cxx_std_20
)

> set(CMAKE_EXECUTABLE_SUFFIX ".js")

Reviewer: Let's cache this.
```suggestion
set(CMAKE_EXECUTABLE_SUFFIX ".js" CACHE STRING "Binary type to be generated by Emscripten")
```

> target_link_options(${BIN_NAME} PRIVATE
>     "-s EXPORTED_FUNCTIONS=['_malloc','_free']"
>     "-s MODULARIZE=1"
>     "-s EXPORT_ES6=1"
>     "-s ENVIRONMENT=worker"
>     "-lembind"
>     "-s ALLOW_MEMORY_GROWTH=1"
>     "-s WASM_BIGINT"
>     "-flto"
>     "--closure 1"
> )

Reviewer: I believe none of the link options were set properly due to the spaces inside each option string. Note that the above syntax generates a list, and when it's being interpreted as a string, we get something like `-s EXPORTED_FUNCTIONS=['_malloc','_free'];-s MODULARIZE=1`. Note how the spaces would cause confusion to the linker (everything after the first space just appear as redundant parameters.)

Sorry for giving you a non-working suggestion in the first place.

Let's do a few improvements:
1. Remove the quotes for the string in the list as those seem unnecessary and only cause confusions.
2. Set up debug / release specific link flags to:
   1. Speed up compile time / runtime.
   2. Set the `ENVIRONMENT` to `node` rather than `worker` so that the generated binary can be run in Node.js for easier debugging.
2. For setting up libraries to be statically linked, we should use `target_link_libraries` because `-lembind` does not seem to work well with `target_link_options`.
4. For the options that take a parameter, remove the space:
   1. For the Boolean options `-s MODULARIZE=1` -> `-sMODULARIZE` or the `"1"` would be intepreted as a string if we do `-sMODULARIZE=1`.
   2. For the string options we simply replace the space with `=`.

```suggestion
if (CMAKE_BUILD_TYPE MATCHES "Release")
    set(EXTRA_LINKER_FLAGS
        -flto
        --closure=1
        -sENVIRONMENT=worker
        CACHE STRING "Extra liner flags for CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}" FORCE
    )
    message("CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}: Enabling optimization linker flags ${EXTRA_LINKER_FLAGS}")
else ()
    set(EXTRA_LINKER_FLAGS
        -sENVIRONMENT=node
        CACHE STRING "Extra liner flags for CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}" FORCE
    )
    message("CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}: Enabling debug flags linker for ${EXTRA_LINKER_FLAGS}")
endif ()
target_link_options(${BIN_NAME} PRIVATE
    ${EXTRA_LINKER_FLAGS}
    -sALLOW_MEMORY_GROWTH
    -sEXPORT_ES6
    -sMODULARIZE
    -sWASM_BIGINT
)
target_link_libraries(${BIN_NAME}
    PRIVATE
    embind
)
```

> target_compile_definitions(${BIN_NAME} PUBLIC SPDLOG_FMT_EXTERNAL=1)

Reviewer: nit: break into multiple lines.
```suggestion
target_compile_definitions(${BIN_NAME}
    PUBLIC
    SPDLOG_FMT_EXTERNAL=1
)
```

> target_include_directories(
>         ${BIN_NAME}
>         PRIVATE
>         ${BOOST_SRC_DIR}
>         src/
>         src/submodules/clp/components/core/src
>         src/submodules/clp/components/core/src/clp
>         src/submodules/clp/components/core/submodules
>         src/submodules/fmt/include
>         src/submodules/spdlog/include
>         src/submodules/zstd/lib
> )

Reviewer: nit:
1. Let's put the target name on the same line of the command.
2. Let's be consistent wit the indentations (4 spaces).
```suggestion
target_include_directories(${BIN_NAME}
    PRIVATE
    ${BOOST_SRC_DIR}
    src/
    src/submodules/clp/components/core/src
    src/submodules/clp/components/core/src/clp
    src/submodules/clp/components/core/submodules
    src/submodules/fmt/include
    src/submodules/spdlog/include
    src/submodules/zstd/lib
)
```

> target_sources(${BIN_NAME}
>         PRIVATE
>         src/bindings.cpp
>         src/ClpIrV1Decoder.cpp
> 
>         src/submodules/clp/components/core/src/clp/ffi/ir_stream/decoding_methods.cpp
>         src/submodules/clp/components/core/src/clp/ir/LogEventDeserializer.cpp
>         src/submodules/clp/components/core/src/clp/ReadOnlyMemoryMappedFile.cpp
>         src/submodules/clp/components/core/src/clp/ReaderInterface.cpp
>         src/submodules/clp/components/core/src/clp/streaming_compression/zstd/Decompressor.cpp
>         src/submodules/clp/components/core/src/clp/TimestampPattern.cpp
> 
>         src/submodules/fmt/src/format.cc
> 
>         src/submodules/zstd/lib/common/entropy_common.c
>         src/submodules/zstd/lib/common/error_private.c
>         src/submodules/zstd/lib/common/fse_decompress.c
>         src/submodules/zstd/lib/common/zstd_trace.c
>         src/submodules/zstd/lib/common/xxhash.c
>         src/submodules/zstd/lib/common/zstd_common.c
>         src/submodules/zstd/lib/decompress/huf_decompress.c
>         src/submodules/zstd/lib/decompress/zstd_ddict.c
>         src/submodules/zstd/lib/decompress/zstd_decompress_block.c
>         src/submodules/zstd/lib/decompress/zstd_decompress.c

Reviewer: nit:
1. Let's be consistent wit the indentations (4 spaces).
```suggestion
target_sources(${BIN_NAME}
    PRIVATE
    src/bindings.cpp
    src/ClpIrV1Decoder.cpp

    src/submodules/clp/components/core/src/clp/ffi/ir_stream/decoding_methods.cpp
    src/submodules/clp/components/core/src/clp/ir/LogEventDeserializer.cpp
    src/submodules/clp/components/core/src/clp/ReadOnlyMemoryMappedFile.cpp
    src/submodules/clp/components/core/src/clp/ReaderInterface.cpp
    src/submodules/clp/components/core/src/clp/streaming_compression/zstd/Decompressor.cpp
    src/submodules/clp/components/core/src/clp/TimestampPattern.cpp

    src/submodules/fmt/src/format.cc

    src/submodules/zstd/lib/common/entropy_common.c
    src/submodules/zstd/lib/common/error_private.c
    src/submodules/zstd/lib/common/fse_decompress.c
    src/submodules/zstd/lib/common/zstd_trace.c
    src/submodules/zstd/lib/common/xxhash.c
    src/submodules/zstd/lib/common/zstd_common.c
    src/submodules/zstd/lib/decompress/huf_decompress.c
    src/submodules/zstd/lib/decompress/zstd_ddict.c
    src/submodules/zstd/lib/decompress/zstd_decompress_block.c
    src/submodules/zstd/lib/decompress/zstd_decompress.c
```

Reviewer: Note we're removing source file `bindings.cpp` as a result of a discussion at https://github.com/y-scope/clp-ffi-js/pull/1/files#r1676350522 . Let's still use "Add suggestion to batch" but we want to remove the reference and commit again.

> )

Reviewer: nit: Add newlines to the end of the files.
```suggestion
)

```

        "Enable/Disable output of compile commands during generation." FORCE)
endif()

# Set the default build type to Release if not specified
if (NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
    set(default_build_type "Release")
    message(STATUS "No build type specified. Setting to '${default_build_type}'.")
    set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE STRING "Choose the type of build." FORCE)
endif ()

# Download and extract Boost
set(BOOST_VERSION "1.85.0" CACHE STRING "Version of Boost library to be used when compiling CLP sources.")
string(REPLACE "." "_" BOOST_ARCHIVE_VERSION_PART ${BOOST_VERSION})

set(BOOST_SRC_DIR "${CMAKE_CURRENT_BINARY_DIR}/boost" CACHE STRING
    "Directory containing Boost sources, fetched from official release if non existent.")
if (NOT EXISTS ${BOOST_SRC_DIR})
    FetchContent_Declare(
        Boost
        URL "https://archives.boost.io/release/${BOOST_VERSION}/source/boost_${BOOST_ARCHIVE_VERSION_PART}.tar.gz"
        SOURCE_DIR ${BOOST_SRC_DIR}
        DOWNLOAD_EXTRACT_TIMESTAMP ON
    )
    message(STATUS "Fetching Boost from official release.")
    FetchContent_MakeAvailable(Boost)

> endif()

Reviewer: ```suggestion
endif ()
```

    )
    message("CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}: Enabling debug flags linker for ${EXTRA_LINKER_FLAGS}")
endif ()
target_link_options(${BIN_NAME} PRIVATE
    ${EXTRA_LINKER_FLAGS}
    -sALLOW_MEMORY_GROWTH
    -sEXPORT_ES6
    -sMODULARIZE
    -sWASM_BIGINT
)
target_link_libraries(${BIN_NAME}
    PRIVATE
    embind
)

target_compile_definitions(${BIN_NAME}
    PUBLIC
    SPDLOG_FMT_EXTERNAL=1
)

target_include_directories(${BIN_NAME}
    PRIVATE
    ${BOOST_SRC_DIR}
    src/
    src/submodules/clp/components/core/src
    src/submodules/clp/components/core/src/clp
    src/submodules/clp/components/core/submodules
    src/submodules/fmt/include
    src/submodules/spdlog/include
    src/submodules/zstd/lib
)

target_sources(${BIN_NAME}
    PRIVATE
    src/ClpIrV1Decoder.cpp

    src/submodules/clp/components/core/src/clp/ffi/ir_stream/decoding_methods.cpp
    src/submodules/clp/components/core/src/clp/ir/LogEventDeserializer.cpp
    src/submodules/clp/components/core/src/clp/ReadOnlyMemoryMappedFile.cpp
    src/submodules/clp/components/core/src/clp/ReaderInterface.cpp
    src/submodules/clp/components/core/src/clp/streaming_compression/zstd/Decompressor.cpp

>     src/submodules/clp/components/core/src/clp/TimestampPattern.cpp

Reviewer 2: As a convention we used in core, we should also include all the header files we used

    src/submodules/fmt/src/format.cc

    src/submodules/zstd/lib/common/entropy_common.c
    src/submodules/zstd/lib/common/error_private.c
    src/submodules/zstd/lib/common/fse_decompress.c

>     src/submodules/zstd/lib/common/zstd_trace.c
>     src/submodules/zstd/lib/common/xxhash.c
>     src/submodules/zstd/lib/common/zstd_common.c

Reviewer 2: ```suggestion
    src/submodules/zstd/lib/common/zstd_common.c
    src/submodules/zstd/lib/common/zstd_trace.c
    src/submodules/zstd/lib/common/xxhash.c
```

    -sMODULARIZE
    -sWASM_BIGINT
)
target_link_libraries(${BIN_NAME}
    PRIVATE
    embind
)

target_compile_definitions(${BIN_NAME}
    PUBLIC
    SPDLOG_FMT_EXTERNAL=1
)

target_include_directories(${BIN_NAME}
    PRIVATE
    ${BOOST_SRC_DIR}
    src/
    src/submodules/clp/components/core/src
    src/submodules/clp/components/core/src/clp
    src/submodules/clp/components/core/submodules
    src/submodules/fmt/include
    src/submodules/spdlog/include
    src/submodules/zstd/lib
)

target_sources(${BIN_NAME}
    PRIVATE
    src/clp_ffi_js/constants.hpp
    src/clp_ffi_js/ClpJsException.hpp
    src/clp_ffi_js/ir/ClpIrV1Decoder.cpp
    src/clp_ffi_js/ir/ClpIrV1Decoder.hpp

    src/submodules/clp/components/core/src/clp/ErrorCode.hpp
    src/submodules/clp/components/core/src/clp/ffi/ir_stream/decoding_methods.cpp
    src/submodules/clp/components/core/src/clp/ffi/ir_stream/decoding_methods.hpp
    src/submodules/clp/components/core/src/clp/ir/LogEvent.hpp
    src/submodules/clp/components/core/src/clp/ir/LogEventDeserializer.cpp
    src/submodules/clp/components/core/src/clp/ir/LogEventDeserializer.hpp
    src/submodules/clp/components/core/src/clp/ir/types.hpp

>     src/submodules/clp/components/core/src/clp/ReadOnlyMemoryMappedFile.cpp

Reviewer 2: Missing headers for this one?

Reviewer: It is not used directly by us currently. I thought we only need to include the ones we directly reference. Let me know if I misunderstood.

Reviewer: Discussed offline with @LinZhihao-723 and @kirkrodrigues to figured out those were needed for CLion to better understand the header sources. Since we found the new language server (ReSharper C++/Rider C++) is better at parsing those. In the cases of a header isn't recognized as part of the project, we can add the `#include` statements in desired source files and the headers will be parsed.

    src/submodules/clp/components/core/src/clp/ReaderInterface.cpp
    src/submodules/clp/components/core/src/clp/streaming_compression/zstd/Decompressor.cpp

>     src/submodules/clp/components/core/src/clp/streaming_compression/zstd/Decompressor.hpp

Reviewer 2: Missing headers for `src/submodules/clp/components/core/src/clp/streaming_compression/Decompressor`?

Reviewer: Similar to [#discussion_r1680236468](#discussion_r1680236468), I'm not sure if we want to include the header if it is not directly used by us.

Reviewer: resolved above

cmake_minimum_required(VERSION 3.16)

include(FetchContent)

if (NOT CMAKE_TOOLCHAIN_FILE)
    set(CMAKE_TOOLCHAIN_FILE
        "${CMAKE_CURRENT_SOURCE_DIR}/build/emsdk/upstream/emscripten/cmake/Modules/Platform/\
Emscripten.cmake"
    )
endif ()

project(ClpIrStreamReader
    LANGUAGES CXX C
)

# Enforce "Unix Makefiles" because "Ninja" does not work with the Emscripten toolchain on CXX_20
# sources.
if (NOT CMAKE_GENERATOR MATCHES "Unix Makefiles")
    message(FATAL_ERROR
        "This project is intended to be built with Unix Makefiles on a Unix system.")
endif ()

# Enable compile commands by default if the generator supports it.

> if (NOT CMAKE_EXPORT_COMPILE_COMMANDS)

Reviewer: I believe this is to respect the setting if specified by the user / IDE. I removed the check and did not find any anomaly in CLion.

> if (NOT CMAKE_EXPORT_COMPILE_COMMANDS)

Reviewer 3: Ah, then I think we should've used `NOT DEFINED CMAKE_EXPORT_COMPILE_COMMANDS` since I think `NOT CMAKE_EXPORT_COMPILE_COMMANDS` will evaluate `CMAKE_EXPORT_COMPILE_COMMANDS` and if it's undefined/OFF, it will run the block (which is the same as not having the condition) and if it's on, then it won't run.

    set(CMAKE_EXPORT_COMPILE_COMMANDS ON CACHE BOOL
        "Enable/Disable output of compile commands during generation." FORCE)
endif ()

# Set the default build type to Release if not specified

> if (NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)

Reviewer: Right, that's not needed for the reason you mentioned.

    set(default_build_type "Release")
    message(STATUS "No build type specified. Setting to '${default_build_type}'.")
    set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE STRING "Choose the type of build." FORCE)
endif ()

# Download and extract Boost
set(BOOST_VERSION "1.85.0" CACHE STRING

>     "Version of Boost library to be used when compiling CLP sources.")

Reviewer 3: ```suggestion
    "Version of Boost library to use.")
```

string(REPLACE "." "_" BOOST_ARCHIVE_VERSION_PART ${BOOST_VERSION})

set(BOOST_SRC_DIR "${CMAKE_CURRENT_BINARY_DIR}/boost" CACHE STRING
    "Directory containing Boost sources, fetched from official release if non existent.")
if (NOT EXISTS ${BOOST_SRC_DIR})
    if (${CMAKE_VERSION} VERSION_GREATER_EQUAL 3.24)
        # Sets DOWNLOAD_EXTRACT_TIMESTAMP=ON in FetchContent_Declare()
        cmake_policy(SET CMP0135 NEW)
    endif ()

    FetchContent_Declare(
        Boost
        URL "https://archives.boost.io/release/${BOOST_VERSION}/source/\

> boost_${BOOST_ARCHIVE_VERSION_PART}.tar.gz"

Reviewer 3: ```suggestion
boost_${BOOST_ARCHIVE_VERSION_PART}.tar.gz"
        URL_MD5 "53aeccc3167909ee770e34469f8dd592"
```

        SOURCE_DIR ${BOOST_SRC_DIR}
    )

>     message(STATUS "Fetching Boost from official release.")

Reviewer 3: ```suggestion
    message(STATUS "Fetching Boost.")
```

    FetchContent_MakeAvailable(Boost)
endif ()

> set(BIN_NAME "ClpIrStreamReader" CACHE STRING

Reviewer 3: Is the plan to create different binaries for different sets of functionality or are we going to create a single binary for the entire library? If it's the latter, perhaps we should call this `ClpFfiJs`?

Reviewer: I was thinking about releasing different binaries for different decoders (when IRv2 comes in) so that we can avoid loading unneeded binaries in the Log Viewer.
We can think more about the separation when IRv2 comes in. I agree `ClpFfijs` is more appropriate for now.

>     "Binary name for the generated .js file and .wasm file of CLP IR V1 Decoder.")

Reviewer 3: ```suggestion
    "Binary name for the generated .js and .wasm files.")
```

add_executable(${BIN_NAME})

target_compile_features(${BIN_NAME}
    PRIVATE
    cxx_std_20
)

> set(CMAKE_EXECUTABLE_SUFFIX ".js" CACHE STRING "Binary type to be generated by Emscripten")

Reviewer 3: ```suggestion
set(CMAKE_EXECUTABLE_SUFFIX ".js" CACHE STRING "Binary type to be generated by Emscripten.")
```

if (CMAKE_BUILD_TYPE MATCHES "Release")
    set(EXTRA_LINKER_FLAGS
        -flto

>         --closure=1

Reviewer 3: Do we need to set `USE_CLOSURE_COMPILER`? (I don't think we do, but just checking.)

Reviewer: I believe that is some legacy interface for requesting use of Closure Compiler. If I read the Emscripten-core code correctly, `USE_CLOSURE_COMPILER` should have been configured when `--closure=1` is set in liner options:
1. The option is parsed here: https://github.com/emscripten-core/emscripten/blob/eb4d36d47f451d0040a0c796606eaae9aa530de3/emcc.py#L1202 .
2. The option is consumed [here](https://github.com/emscripten-core/emscripten/blob/eb4d36d47f451d0040a0c796606eaae9aa530de3/tools/link.py#L953) to set `settings.USE_CLOSURE_COMPILER`.

        -sENVIRONMENT=worker

>         CACHE STRING "Extra liner flags for CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}" FORCE

Reviewer 3: ```suggestion
        CACHE STRING "Extra linker flags for CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}" FORCE
```

    )
    message("CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}: Enabling optimization linker flags \
${EXTRA_LINKER_FLAGS}")
else ()
    set(EXTRA_LINKER_FLAGS
        -sENVIRONMENT=node

>         CACHE STRING "Extra liner flags for CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}" FORCE

Reviewer 3: ```suggestion
        CACHE STRING "Extra linker flags for CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}" FORCE
```

    )

>     message("CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}: Enabling debug flags linker for \

Reviewer 3: Can we replace this and the other message with a single message after the if-else:

```
message("CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}: Extra linker flags: ${EXTRA_LINKER_FLAGS}")
```

${EXTRA_LINKER_FLAGS}")

> endif ()

Reviewer 3: Since we're forcibly setting the flags, we don't need to make it a cache variable, right?

Reviewer: Right, assuming we will toggle the CMAKE_BUILD_TYPE with the same build directory. Removed both `CACHE` and `FORCE` now.

)

target_sources(${BIN_NAME}
    PRIVATE
    src/clp_ffi_js/ir/StreamReader.cpp

    src/submodules/clp/components/core/src/clp/ffi/ir_stream/decoding_methods.cpp
    src/submodules/clp/components/core/src/clp/ir/EncodedTextAst.cpp
    src/submodules/clp/components/core/src/clp/ir/LogEventDeserializer.cpp
    src/submodules/clp/components/core/src/clp/ReadOnlyMemoryMappedFile.cpp
    src/submodules/clp/components/core/src/clp/ReaderInterface.cpp
    src/submodules/clp/components/core/src/clp/streaming_compression/zstd/Decompressor.cpp
    src/submodules/clp/components/core/src/clp/TimestampPattern.cpp

    src/submodules/fmt/src/format.cc

    src/submodules/zstd/lib/common/entropy_common.c
    src/submodules/zstd/lib/common/error_private.c
    src/submodules/zstd/lib/common/fse_decompress.c
    src/submodules/zstd/lib/common/zstd_common.c
    src/submodules/zstd/lib/common/zstd_trace.c
    src/submodules/zstd/lib/common/xxhash.c
    src/submodules/zstd/lib/decompress/huf_decompress.c
    src/submodules/zstd/lib/decompress/zstd_ddict.c
    src/submodules/zstd/lib/decompress/zstd_decompress_block.c
    src/submodules/zstd/lib/decompress/zstd_decompress.c

> )

Reviewer 3: Can we split these groups into separate variables so that there's less chance of mixing them up when IDEs add to the list?

cmake_minimum_required(VERSION 3.22)

> 

Reviewer: Let's place `include()`s at the top of the file
```suggestion
include(FetchContent)

```

Reviewer: @xx12345798 As we discussed in the sync-up, it will be more ideal if we checkout the other dependencies (e.g., zstd, fmt, clp, spdlog) within the CMakeLists.txt file with the FetchContent macros. On the other hand, since we wish to get the PR merged early, and the current setup is working nicely, we can do that in another cleanup PR or when we have the need to upgrade the dependencies.

> project(ClpIrV1Decoder)

Reviewer: ```suggestion
project(ClpIrV1Decoder
    LANGUAGES CXX C
)
```

> 

Reviewer: ```suggestion

# Enforce "Unix Makefiles" because "Ninja" does not work with the Emscripten toolchain on CXX_20 sources.
if(NOT CMAKE_GENERATOR MATCHES "Unix Makefiles")
    message(FATAL_ERROR "This project is intended to be built with Unix Makefiles on a Unix system.")
endif()

# Enable compile commands by default if the generator supports it.
if (NOT CMAKE_EXPORT_COMPILE_COMMANDS AND CMAKE_GENERATOR MATCHES "Unix Makefiles")
    set(CMAKE_EXPORT_COMPILE_COMMANDS ON CACHE BOOL
        "Enable/Disable output of compile commands during generation." FORCE)
endif()

# Set the default build type to Release if not specified
if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
    message(STATUS "Setting build type to 'Release' as none was specified.")
    set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Choose the type of build (Debug, Release, RelWithDebInfo, MinSizeRel)." FORCE)
endif()
```

# Download and extract Boost

> set(BOOST_VERSION "1.85.0")

Reviewer: Let's enable caching on the variable and also add a docstring for it.
```suggestion
set(BOOST_VERSION "1.85.0" CACHE STRING "Version of Boost library to be used when compiling CLP sources.")
```

string(REPLACE "." "_" BOOST_ARCHIVE_VERSION_PART ${BOOST_VERSION})

> include(ExternalProject)
> ExternalProject_Add(
>         Boost
>         URL "https://archives.boost.io/release/${BOOST_VERSION}/source/boost_${BOOST_ARCHIVE_VERSION_PART}.tar.gz"
>         PREFIX "${CMAKE_BINARY_DIR}/boost"
>         CONFIGURE_COMMAND ""
>         BUILD_COMMAND ""
>         INSTALL_COMMAND ""
>         LOG_DOWNLOAD ON
> )
> ExternalProject_Get_Property(Boost SOURCE_DIR)
> set(BOOST_EXTRACT_DIR "${SOURCE_DIR}")

Reviewer: Looking online it seems a better alternative is to use `FetchContent_Declare` + `FetchContent_MakeAvailable`. 

Let's place `include()` at the top of the file just as how we include headers in source files in source codes. (See the top for the suggested change. )

```suggestion
set(BOOST_SRC_DIR "${CMAKE_CURRENT_SOURCE_DIR}/boost" CACHE STRING
    "Directory containing Boost sources, fetched from official release if non existent.")
if (NOT EXISTS ${BOOST_SRC_DIR})
    FetchContent_Declare(
        Boost
        URL "https://archives.boost.io/release/${BOOST_VERSION}/source/boost_${BOOST_ARCHIVE_VERSION_PART}.tar.gz"
        SOURCE_DIR ${BOOST_SRC_DIR}
    )
    message(STATUS "Fetching Boost from official release.")
    FetchContent_MakeAvailable(Boost)
endif()
```

> 

Reviewer: Let's separate the target definition and source file additions from a single `add_executable` from below.

```suggestion

set(BIN_NAME "ClpIrV1Decoder" CACHE STRING "Binary name for the generated .js file and .wasm file of CLP IR V1 Decoder.")
add_executable(${BIN_NAME})
```

> set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++23")
> set(CMAKE_CXX_STANDARD 23)

Reviewer: Sorry for sharing you some outdated config in the first place. 

It seems the norm these days is to use `target_compile_features(targetName visibility cxx_std_XX)` because it sets the language standard per target, assuming we will add more targets into this CMakeLists.txt in the future like 
a `ClpIrV2Decoder`.

Let's also drop the CXX standard to 20 (instead of 23) to stay consistent with clp and our other ffi repos. 

```suggestion
target_compile_features(${BIN_NAME}
    PRIVATE
    cxx_std_20
)
```

set(CMAKE_EXECUTABLE_SUFFIX ".js")

> add_executable(
>         ClpIrV1Decoder
> 

Reviewer: See above for the comment to separate `add_executable`.
```suggestion
target_sources(${BIN_NAME} PRIVATE
```

Reviewer: After applying all suggestions, let's place this `target_sources` block at the end of the place so the sources are specified after all compilation options and include directory specifications.

        src/ClpIrV1Decoder.cpp

        src/submodules/clp/components/core/src/clp/ffi/ir_stream/decoding_methods.cpp
        src/submodules/clp/components/core/src/clp/ir/LogEventDeserializer.cpp
        src/submodules/clp/components/core/src/clp/ReadOnlyMemoryMappedFile.cpp
        src/submodules/clp/components/core/src/clp/ReaderInterface.cpp
        src/submodules/clp/components/core/src/clp/streaming_compression/zstd/Decompressor.cpp
        src/submodules/clp/components/core/src/clp/TimestampPattern.cpp

        src/submodules/fmt/src/format.cc

        src/submodules/zstd/lib/common/entropy_common.c
        src/submodules/zstd/lib/common/error_private.c
        src/submodules/zstd/lib/common/fse_decompress.c
        src/submodules/zstd/lib/common/zstd_trace.c
        src/submodules/zstd/lib/common/xxhash.c
        src/submodules/zstd/lib/common/zstd_common.c
        src/submodules/zstd/lib/decompress/huf_decompress.c
        src/submodules/zstd/lib/decompress/zstd_ddict.c
        src/submodules/zstd/lib/decompress/zstd_decompress_block.c
        src/submodules/zstd/lib/decompress/zstd_decompress.c
)

> set_target_properties(ClpIrV1Decoder PROPERTIES COMPILE_FLAGS "-O3")

Reviewer: Now the optimization-related compile flags are controlled by `CMAKE_BUILD_TYPE` (see how we also specifying the default to be "Release" above), we do not / should not directly specify the compilation flags here.
```suggestion
```

> set_target_properties(ClpIrV1Decoder PROPERTIES LINK_FLAGS    "-s EXPORTED_FUNCTIONS=['_malloc','_free'] -s MODULARIZE=1 -s EXPORT_ES6=1 -s ENVIRONMENT=worker -lembind -sALLOW_MEMORY_GROWTH=1 -sWASM_BIGINT -flto -O3 --closure 1")

Reviewer: Let's use this new syntax for specifying link flags. Also note the `-O3` flag has been removed since it should also be automatically added if `CMAKE_BUILD_TYPE` is "Release".
```suggestion
target_link_options(${BIN_NAME} PRIVATE
    "-s EXPORTED_FUNCTIONS=['_malloc','_free']"
    "-s MODULARIZE=1"
    "-s EXPORT_ES6=1"
    "-s ENVIRONMENT=worker"
    "-lembind"
    "-s ALLOW_MEMORY_GROWTH=1"
    "-s WASM_BIGINT"
    "-flto"
    "--closure 1"
)
```

target_compile_definitions(ClpIrV1Decoder PUBLIC SPDLOG_FMT_EXTERNAL=1)

target_include_directories(

>         ClpIrV1Decoder

Reviewer: ```suggestion
        ${BIN_NAME}
```

        PRIVATE

>         ${BOOST_EXTRACT_DIR}

Reviewer: ```suggestion
        ${BOOST_SRC_DIR}
```
